### PR TITLE
Resolves segfaults if state for context not managed explicitly

### PR DESCRIPTION
**Pull request by badisa**
* Returns BoundPotentials wrapped in shared_ptr
* Returns Integrators wrapped in shared_ptr
* Returns MonteCarloBarostat wrapped in shared_ptr
* Ownership semantics are cleaner, and will make some code in local MD cleaner

### CODE REVIEW

**Comment by maxentile** [comment]
Nice!

Should allow me to delete blocks like

```python
...
ctxt = custom_ops.Context(coords, velocities, box, intg_impl, all_impls, barostat_impl)

# otherwise code that returns only `ctxt` will segfault...
ctxt._all_impls = all_impls
ctxt._barostat = barostat
ctxt._barostat_impl = barostat_impl
ctxt._intg = intg
ctxt._intg_impl = intg_impl
return ctxt
```

from experiment scripts and other places where it was useful to allow a function to return a `custom_ops.Context`.

**Comment by maxentile** [comment]
Another site in `tests/` that can be updated to exercise this fix is `return intg, context` in `test_velocity_verlet_integrator.py`  https://github.com/proteneer/timemachine/pull/789#discussion_r941822300

**Review by maxentile — APPROVED** [review]
Nice!

**Comment by badisa** [comment]
> Another site in `tests/` that can be updated to exercise this fix is `return intg, context` in `test_velocity_verlet_integrator.py` [#789 (comment)](https://github.com/proteneer/timemachine/pull/789#discussion_r941822300)

Fixed in [87e6cb7](https://github.com/proteneer/timemachine/pull/972/commits/87e6cb765067db70f0780be83e1e66c38c721049)

File: tests/test_md.py [hunk 1]
```diff
@@ -529,6 +530,56 @@ def test_multiple_steps_local_consistency(self):
...
+            intg = LangevinIntegrator(temperature, dt, friction, masses, seed)
+
+            barostat_impl = None
+            if barostat_interval > 0:
+                group_idxs = get_group_indices(get_bond_list(unbound_potentials[0]))
+
+                barostat = MonteCarloBarostat(coords.shape[0], pressure, temperature, group_idxs, 1, seed)
+                barostat_impl = barostat.impl(bps)
+
+            return custom_ops.Context(coords, v0, box, intg.impl(), bps, barostat=barostat_impl)
```
Comment (mcwitt) [inline tests/test_md.py:564]: Idea for strengthening test: return weakrefs (e.g. `intg` and `barostat`) from this function and use to assert that they're cleaned up after `del ctxt; gc.collect()` below

**Review by mcwitt — APPROVED** [review]
Looks great!

**Pull request merged**